Add private_class_method to fix lint warnings

AndyStabler opened the pull request:
Fixes #12035 

Adds `private_class_method` to class methods that are private.

I think I've caught all of the relevant warnings, but I'll take another look tomorrow night!

jrafanie commented:
@AndyStabler Some tests are failing. Can you "undo" each of the changes where it's complaining about a private method and instead comment something along the lines of:

```ruby
# FIXME: The intention was to make this method private but tests or code called from the tests are expecting it to be public.
```

We can deal with the "easy" ones in this PR and fix the harder ones later.  Thanks again for your contribution!

jrafanie commented:
@AndyStabler Also, don't worry about fixing existing code complaints reported by @miq-bot, only fix things related to your changes in this PR.  👍

AndyStabler commented:
Cool, thanks for taking a look, @jrafanie. I'll fix up the remaining broken tests tomorrow and clean up the code to add a space before and after `protected` tomorrow.

jrafanie commented on gems/pending/disk/modules/MSCommon.rb:
```diff
@@ -219,134 +219,136 @@ def self.getFooter(file, skip_check = false)
-    unless skip_check
-      headerCsum = checksum(buf, 36)
-      raise "Header checksum doesn't match: got 0x#{'%04x' % headerCsum}, s/b 0x#{'%04x' % @header['checksum']}" if headerCsum != header['checksum']
+  class << self
```
@AndyStabler thanks for going through all of these... for this one... we prefer using private_class_method over `class << self` with a `private` section for readability.  It's easy to keep adding to the singleton class and not realizing what context you're in, especially when there's so many pages of singleton methods.  Feel free to either do this file or skip it.  We can finish up the annoying ones.  Your contribution is greatly appreciated!

AndyStabler replied:
Ah, ok- here you go: 9b442b2

jrafanie commented on gems/pending/util/miq-password.rb:
```diff
@@ -174,26 +174,28 @@ def self.generate_symmetric(filename = nil)
-    else
-      params = YAML.load_file(filename)
-      CryptString.new(nil, params[:algorithm], params[:key], params[:iv])
+  class << self
```
Same as above RE: using private_class_method, sorry 😢

AndyStabler replied:
No problem 🙂 7dc3815

jrafanie commented on gems/pending/util/mount/miq_generic_mount_session.rb:
```diff
@@ -451,14 +451,18 @@ def with_mounted_exception_handling
-    else
-      raise "platform not supported"
+  class << self
+    protected
```
Same as above RE: private_class method.  Why protected?

AndyStabler replied:
I wasn't quite sure how to tackle this. The class methods were originally beneath [`protected`](https://github.com/ManageIQ/manageiq/pull/12041/files/980057de1f7f59e96d2d9fcc9f392238f8ee43b5#diff-8b37bf510f735b78222889d7d5763cd2R427), so I created the singleton class and added `protected` inside it. Happy to move these to `private_class_method`s though if that makes sense.

AndyStabler replied:
Here you go: 0eeb03f

AndyStabler commented:
I'll get rid of the sneaky character at https://github.com/AndyStabler/manageiq/blob/0eeb03f2d4ba797d82b2d8059972399d38feaf07/app/models/rss_feed.rb#L80 tonight. Then we should be good for another review, @jrafanie 🙂

AndyStabler commented:
Alrighty, @jrafanie could you take another look at this, please?

jrafanie commented:
Thanks @AndyStabler, I'll take a look shortly.  Thanks again!

jrafanie commented on gems/pending/util/miq-password.rb:
```diff
@@ -191,10 +189,12 @@ def self.ez_load(filename, recent = true)
       CryptString.new(nil, params[:algorithm], params[:key], params[:iv])
     end
   end
+  private_class_method :ez_load
```
@kbrock do you have concern about this going from protected to private?  the tests don't fail.  😨

jrafanie replied:
Note, there's no such thing as a protected class method so it was always public before... so it's going from public to private.

kbrock replied:
`keys` is tested a bunch. So if I'm pretty sure a private `ez_load` is still working here. :shipit:

jrafanie replied:
Thanks @kbrock

jrafanie commented on app/models/manageiq/providers/openstack/cloud_manager/provision_workflow.rb:
```diff
@@ -91,4 +91,5 @@ def dialog_name_from_automate(message = 'get_dialog_name')
   def self.provider_model
     ManageIQ::Providers::Openstack::CloudManager
   end
+  private_class_method :provider_model
```
I think we have to undo this one.  It probably is meant to be public.  cc @Ladas 

```
app/models/manageiq/providers/azure/cloud_manager/provision_workflow.rb:  def self.provider_model
app/models/manageiq/providers/azure/cloud_manager/provision_workflow.rb:  private_class_method :provider_model
app/models/manageiq/providers/google/cloud_manager/provision_workflow.rb:  def self.provider_model
app/models/manageiq/providers/google/cloud_manager/provision_workflow.rb:  private_class_method :provider_model
app/models/manageiq/providers/microsoft/infra_manager/provision_workflow.rb:  def self.provider_model
app/models/manageiq/providers/openstack/cloud_manager/provision_workflow.rb:  def self.provider_model
app/models/manageiq/providers/openstack/cloud_manager/provision_workflow.rb:  private_class_method :provider_model
app/models/manageiq/providers/redhat/infra_manager/provision_workflow.rb:  def self.provider_model
app/models/manageiq/providers/vmware/infra_manager/provision_workflow.rb:  def self.provider_model
app/models/miq_provision_virt_workflow.rb:    return ["vms.template = ? AND vms.ems_id IS NOT NULL", true] unless self.class.respond_to?(:provider_model)
app/models/miq_provision_virt_workflow.rb:    ["vms.template = ? AND vms.ems_id in (?)", true, self.class.provider_model.pluck(:id)]
```

jrafanie replied:
i think

Ladas replied:
yes, needs to be public, it's called from app/models/miq_provision_virt_workflow.rb

jrafanie replied:
thanks @Ladas

jrafanie commented on app/models/manageiq/providers/foreman/provider.rb:
```diff
@@ -74,4 +74,5 @@ def ensure_managers
   def self.refresh_ems(provider_ids)
     EmsRefresh.queue_refresh(Array.wrap(provider_ids).collect { |id| [base_class, id] })
   end
+  private_class_method :refresh_ems
```
@bdunne does this look ok? I don't see anything calling self.class.refresh_ems or similar, which would indicate it's truly not private.

bdunne replied:
@jrafanie The only thing I'm concerned about is this method being queued.

jrafanie commented on app/models/log_file.rb:
```diff
@@ -267,4 +267,5 @@ def self._request_logs(options)
     _log.info("#{log_header} #{msg}")
     task.update_status("Queued", "Ok", msg)
   end
+  private_class_method :_request_logs
```
I think this is meant to be public and needs to be moved before the private keyword.  See above, we put something on the queue, to be run:

```ruby
       MiqQueue.put(
         :class_name   => name,
         :method_name  => "_request_logs",
         :args         => [options],
         :zone         => zone,
         :miq_callback => cb,
         :msg_timeout  => LOG_REQUEST_TIMEOUT,
         :priority     => MiqQueue::HIGH_PRIORITY
       )
```

So, this is a public method with a bad name.

jrafanie commented on app/models/classification.rb:
```diff
@@ -521,6 +524,7 @@ def self.tag2human(tag)
 
     "#{cname}: #{ename}"
   end
+  private_class_method :tag2human
```
This looks to be a public method that should be before the private keyword.

```
app/controllers/miq_policy_controller/miq_actions.rb:    @tag_selected = Classification.tag2human(tag_name)
app/models/classification.rb:  def self.tag2human(tag)
app/models/classification.rb:  private_class_method :tag2human
app/views/miq_policy/_action_details.html.haml:                = h(Classification.tag2human(@action.options[:tags].first))
```

```ruby
irb(main):001:0> Classification.tag2human("blah")
NoMethodError: PostgreSQLAdapter#log_after_checkout, connection_pool: size: 5, connections: 1, in use: 1
private method `tag2human' called for #<Class:0x007ffc62ced360>

```

kbrock replied:
good find. Guess we need a test for `Classification.tag2human` (not part of this PR)

jrafanie commented on lib/miq_automation_engine/models/miq_ae_domain.rb:
```diff
@@ -170,11 +170,16 @@ def squeeze_priorities
   def self.any_enabled?
     MiqAeDomain.enabled.count > 0
   end
+  private_class_method :any_enabled?
```
we call this with a receiver in the toolbar_builder.rb, so this method should be moved before the private keyword in a followup PR, so drop this change.

```
git grep any_enabled
app/helpers/application_helper/toolbar_builder.rb:    MiqAeDomain.any_enabled?
lib/miq_automation_engine/models/miq_ae_domain.rb:  def self.any_enabled?
lib/miq_automation_engine/models/miq_ae_domain.rb:  private_class_method :any_enabled?
```

jrafanie commented on gems/pending/util/xml/miq_rexml.rb:
```diff
@@ -106,6 +105,7 @@ def self.findElementInt(paths, ele)
       return ele
     end
   end
+  private_class_method :findElementInt
```
Unfortunately, I think we have to leave this as it was...   Maybe just drop this change or add a comment.  We really have to dump all this rexml stuff, but not today sadly.  😢   cc @gmcculloug 

I can't tell if this is being called with a receiver:

```
git grep findElementInt
gems/pending/util/xml/miq_rexml.rb:    findElementInt(path.tr("\\", "/").split("/"), element)
gems/pending/util/xml/miq_rexml.rb:  def self.findElementInt(paths, ele)
gems/pending/util/xml/miq_rexml.rb:          return findElementInt(paths, e)
gems/pending/util/xml/miq_rexml.rb:  private_class_method :findElementInt
gems/pending/util/xml/xml_hash.rb:    def self.findElementInt(paths, ele)
gems/pending/util/xml/xml_hash.rb:            return findElementInt(paths, e)
gems/pending/util/xml/xml_utils.rb:    return XmlHash::XmhHelpers.findElementInt(path_fix_up, element) if element.kind_of?(Hash)
gems/pending/util/xml/xml_utils.rb:    findElementInt(path_fix_up, element)
gems/pending/util/xml/xml_utils.rb:  def self.findElementInt(paths, ele)
gems/pending/util/xml/xml_utils.rb:          return findElementInt(paths, e)
```

jrafanie commented on gems/pending/util/xml/miq_rexml.rb:
```diff
@@ -83,6 +81,7 @@ def self.findRegElementInt(paths, ele)
       return ele
     end
   end
+  private_class_method :findRegElementInt
```
Same as below... 😢

jrafanie commented on gems/pending/metadata/VmConfig/VmConfig.rb:
```diff
@@ -850,6 +850,7 @@ def self.to_h(filename)
     end
     dataHash
   end
+  private_class_method :to_h
```
This was definitely a mistake originally, `to_h` is a public method.  If we ever intended it to be private, we were wrong.  Please remove this change and do it in a followup PR or I'll do it, just say the word.

jrafanie commented on gems/pending/disk/modules/MSCommon.rb:
```diff
@@ -349,4 +363,5 @@ def self.checksum(buf, skip_offset)
     # GRRRRR - convert to actual 32-bits.
     [~csum].pack('L').unpack('L')[0]
   end
+  private_class_method :checksum
```
@jerryk55 can you review these changes.  Are these private methods in this file?  The gist is that class/singleton methods that follow the private keyword aren't actually private.  You can't make them private in that way.  You need to use our preferred way, `private_class_method`, or make them private in the singleton class.

jrafanie replied:
Unping, I went through and git grep'd around and nothing is calling these private class methods with an explicit  receiver (not even self).  I'm good with merging this as euwe/no (e release) so we have time to test out these changes for the F-release.

jrafanie commented:
@AndyStabler I made several comments, after we get some feedback from others, can you undo some of the changes and squash your commits?  Thank you.

AndyStabler commented:
Brilliant, thanks for the feedback, @jrafanie. I'll take a look at this tomorrow night 🙂

kbrock commented on app/models/manageiq/providers/base_manager/refresher.rb:
```diff
@@ -25,6 +25,8 @@ def refresher_options
       private
 
+      # FIXME: The intention was to make this method private but tests or code called
+      # from the tests are expecting it to be public.
```
I don't know that the intent is for this to be private.

jrafanie replied:
I"m ok with dropping this comment and moving this method before the private keyword in a followup PR.

jrafanie replied:
Thanks @kbrock, I agree, let's remove this comment now and move this method before private in a followup PR

kbrock commented on app/models/manageiq/providers/openstack/cloud_manager/cloud_tenant.rb:
```diff
@@ -47,6 +47,7 @@ def self.connection_options
     connection_options = {:service => "Identity", :openstack_endpoint_type => 'adminURL'}
     connection_options
   end
+  private_class_method :connection_options
```
Having trouble parsing. Do you mean "so we can't use private"?

jrafanie replied:
yes:

```ruby
   def connection_options
      self.class.connection_options
   end
```

Because we have a receiver, we can't make this method private.  We need to move this method before the private keyword in a followup PR.

movitto commented on gems/pending/VolumeManager/LVM/scanner.rb:
```diff
@@ -149,8 +150,10 @@ def self.readRaw(d, rlh)
   def self.readStruct(d, struct)
     OpenStruct.new(struct.decode(d.read(struct.size)))
   end # def self.readStruct
+  private_class_method :readStruct
```
@jrafanie  @AndyStabler Looks good. I verified the affected methods in this class are just called internally, applied the changes to this module and successfully ran a smart state locally.

:+1:

jrafanie replied:
Awesome, thanks @movitto 🙇

AndyStabler commented:
OK, I've reverted the changes you highlighted @jrafanie. I've also squashed my commits. I'll create a follow up PR to make all of the private methods you mentioned public. :+1:

jrafanie commented:
@AndyStabler I restarted the travis build... it looks like you were victim of a sporadic test failure.

![image](https://cloud.githubusercontent.com/assets/19339/19823605/811c22e2-9d37-11e6-9fa4-ab89230d9930.png)

Opened #12288 to fix that problem... hopefully your's will pass

AndyStabler commented:
@jrafanie Thanks for restarting the build and for the review 😄 I've put together a [PR](https://github.com/ManageIQ/manageiq/pull/12321) to fix the private methods that should be public.

AndyStabler commented:
Cool! Here's the [revert commit](ca5f765), @jrafanie. I also made `refresh_ems` public over at https://github.com/ManageIQ/manageiq/pull/12321

jrafanie commented:
Looks good @AndyStabler, we're in blockers only on our euwe release branch so I'm holding off on merging this right now to avoid master causing backport conflicts.  Will merge soon.  Thanks!

jrafanie commented:
@AndyStabler Can you squash the "undo" commit?  Thanks!

AndyStabler commented:
@jrafanie sure, that should be squashed now :+1:

jrafanie commented:
Thanks @AndyStabler

bdunne commented:
@jrafanie Should this be rebased and merged?
